Use endpoint instead of hardcoded JSON

ywkim312 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This PR changes the frontend to use the json obtained from the end point instead of saved json file

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
You have to be either in NCSANet or vpn connected to ncsa
In your, command shell, create env variable by export APP_ENV="development"
Then run the frontend. If you see the All Programs map and its tooltip when you hove your mouse over the map, then the PR works well. Before, the values in tooltips came from the saved json but in this PR, it requests the data from the end point and use the values obtained from the end point

Currently, it is using the localhost only but you can change the information in app.config.js and this app.config.js should be more carefully designed and added when the pdl-api gets deployed.

## Screenshots (if appropriate):
![image](https://user-images.githubusercontent.com/12547573/216707748-adc2bef7-6418-412c-b162-b28fa9de5dd9.png)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation and CHANGELOG accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

ywkim312 commented:
> sorry I'm definitely new to this.. @ywkim312 I use `NCSA SSLVPN` but still cannot fetch the data. Am I wrong in choosing the VPN?

you have to set the  environmental variable APP_ENV="development" when you run the npm

pengyin-shan commented:
> > sorry I'm definitely new to this.. @ywkim312 I use `NCSA SSLVPN` but still cannot fetch the data. Am I wrong in choosing the VPN?
> 
> you have to set the environmental variable APP_ENV="development" when you run the npm

Like this right?

<img width="444" alt="Screenshot 2023-03-20 at 2 16 42 PM" src="https://user-images.githubusercontent.com/92752107/226443172-b66a32ae-4bf1-4dfc-ac4c-4168cf2997b8.png">

ywkim312 commented:
> > > sorry I'm definitely new to this.. @ywkim312 I use `NCSA SSLVPN` but still cannot fetch the data. Am I wrong in choosing the VPN?
> > 
> > 
> > you have to set the environmental variable APP_ENV="development" when you run the npm
> 
> Like this right?
> 
> <img alt="Screenshot 2023-03-20 at 2 16 42 PM" width="444" src="https://user-images.githubusercontent.com/92752107/226443172-b66a32ae-4bf1-4dfc-ac4c-4168cf2997b8.png">

you can try 
```
set APP_ENV="development"
```

pengyin-shan commented:
> > > > sorry I'm definitely new to this.. @ywkim312 I use `NCSA SSLVPN` but still cannot fetch the data. Am I wrong in choosing the VPN?
> > > 
> > > 
> > > you have to set the environmental variable APP_ENV="development" when you run the npm
> > 
> > 
> > Like this right?
> > <img alt="Screenshot 2023-03-20 at 2 16 42 PM" width="444" src="https://user-images.githubusercontent.com/92752107/226443172-b66a32ae-4bf1-4dfc-ac4c-4168cf2997b8.png">
> 
> you can try
> 
> ```
> set APP_ENV="development"
> ```

Yes, I set up environment by exporting `APP_ENV="development"` and getting `console.log` output `development` environment, which is correct. 

I think the problem is still in VPN side, because I obviously cannot access data from my machine. Visiting `https://api.policydesignlab-dev.ncsa.illinois.edu/pdl/statecodes` returns `Forbidden` also:
<img width="1710" alt="Screenshot 2023-03-20 at 2 42 48 PM" src="https://user-images.githubusercontent.com/92752107/226448670-f6b63431-22e9-4102-a498-bd7f6634ebfc.png">

It is `sslvpn.ncsa.illinois.edu` right?

ywkim312 commented:
> > > > > sorry I'm definitely new to this.. @ywkim312 I use `NCSA SSLVPN` but still cannot fetch the data. Am I wrong in choosing the VPN?
> > > > 
> > > > 
> > > > you have to set the environmental variable APP_ENV="development" when you run the npm
> > > 
> > > 
> > > Like this right?
> > > <img alt="Screenshot 2023-03-20 at 2 16 42 PM" width="444" src="https://user-images.githubusercontent.com/92752107/226443172-b66a32ae-4bf1-4dfc-ac4c-4168cf2997b8.png">
> > 
> > 
> > you can try
> > ```
> > set APP_ENV="development"
> > ```
> 
> Yes, I set up environment by exporting `APP_ENV="development"` and getting `console.log` output `development` environment, which is correct.
> 
> I think the problem is still in VPN side, because I obviously cannot access data from my machine. Visiting `https://api.policydesignlab-dev.ncsa.illinois.edu/pdl/statecodes` returns `Forbidden` also: <img alt="Screenshot 2023-03-20 at 2 42 48 PM" width="1710" src="https://user-images.githubusercontent.com/92752107/226448670-f6b63431-22e9-4102-a498-bd7f6634ebfc.png">
> 
> It is `sslvpn.ncsa.illinois.edu` right?

I think that you have to use cisco anyconnect https://answers.uillinois.edu/illinois/page.php?id=98773

pengyin-shan commented:
> > > > > > sorry I'm definitely new to this.. @ywkim312 I use `NCSA SSLVPN` but still cannot fetch the data. Am I wrong in choosing the VPN?
> > > > > 
> > > > > 
> > > > > you have to set the environmental variable APP_ENV="development" when you run the npm
> > > > 
> > > > 
> > > > Like this right?
> > > > <img alt="Screenshot 2023-03-20 at 2 16 42 PM" width="444" src="https://user-images.githubusercontent.com/92752107/226443172-b66a32ae-4bf1-4dfc-ac4c-4168cf2997b8.png">
> > > 
> > > 
> > > you can try
> > > ```
> > > set APP_ENV="development"
> > > ```
> > 
> > 
> > Yes, I set up environment by exporting `APP_ENV="development"` and getting `console.log` output `development` environment, which is correct.
> > I think the problem is still in VPN side, because I obviously cannot access data from my machine. Visiting `https://api.policydesignlab-dev.ncsa.illinois.edu/pdl/statecodes` returns `Forbidden` also: <img alt="Screenshot 2023-03-20 at 2 42 48 PM" width="1710" src="https://user-images.githubusercontent.com/92752107/226448670-f6b63431-22e9-4102-a498-bd7f6634ebfc.png">
> > It is `sslvpn.ncsa.illinois.edu` right?
> 
> I think that you have to use cisco anyconnect https://answers.uillinois.edu/illinois/page.php?id=98773

Ah, yes I am..here is the setting:
<img width="425" alt="Screenshot 2023-03-20 at 2 48 54 PM" src="https://user-images.githubusercontent.com/92752107/226450014-80443c03-193a-4d36-afa7-220277c064e5.png">

ywkim312 commented:
> > > > > > > sorry I'm definitely new to this.. @ywkim312 I use `NCSA SSLVPN` but still cannot fetch the data. Am I wrong in choosing the VPN?
> > > > > > 
> > > > > > 
> > > > > > you have to set the environmental variable APP_ENV="development" when you run the npm
> > > > > 
> > > > > 
> > > > > Like this right?
> > > > > <img alt="Screenshot 2023-03-20 at 2 16 42 PM" width="444" src="https://user-images.githubusercontent.com/92752107/226443172-b66a32ae-4bf1-4dfc-ac4c-4168cf2997b8.png">
> > > > 
> > > > 
> > > > you can try
> > > > ```
> > > > set APP_ENV="development"
> > > > ```
> > > 
> > > 
> > > Yes, I set up environment by exporting `APP_ENV="development"` and getting `console.log` output `development` environment, which is correct.
> > > I think the problem is still in VPN side, because I obviously cannot access data from my machine. Visiting `https://api.policydesignlab-dev.ncsa.illinois.edu/pdl/statecodes` returns `Forbidden` also: <img alt="Screenshot 2023-03-20 at 2 42 48 PM" width="1710" src="https://user-images.githubusercontent.com/92752107/226448670-f6b63431-22e9-4102-a498-bd7f6634ebfc.png">
> > > It is `sslvpn.ncsa.illinois.edu` right?
> > 
> > 
> > I think that you have to use cisco anyconnect https://answers.uillinois.edu/illinois/page.php?id=98773
> 
> Ah, yes I am..here is the setting: <img alt="Screenshot 2023-03-20 at 2 48 54 PM" width="425" src="https://user-images.githubusercontent.com/92752107/226450014-80443c03-193a-4d36-afa7-220277c064e5.png">

hmmm that is super weird. I just tested and it worked well

ywkim312 commented:
> > > > > > > sorry I'm definitely new to this.. @ywkim312 I use `NCSA SSLVPN` but still cannot fetch the data. Am I wrong in choosing the VPN?
> > > > > > 
> > > > > > 
> > > > > > you have to set the environmental variable APP_ENV="development" when you run the npm
> > > > > 
> > > > > 
> > > > > Like this right?
> > > > > <img alt="Screenshot 2023-03-20 at 2 16 42 PM" width="444" src="https://user-images.githubusercontent.com/92752107/226443172-b66a32ae-4bf1-4dfc-ac4c-4168cf2997b8.png">
> > > > 
> > > > 
> > > > you can try
> > > > ```
> > > > set APP_ENV="development"
> > > > ```
> > > 
> > > 
> > > Yes, I set up environment by exporting `APP_ENV="development"` and getting `console.log` output `development` environment, which is correct.
> > > I think the problem is still in VPN side, because I obviously cannot access data from my machine. Visiting `https://api.policydesignlab-dev.ncsa.illinois.edu/pdl/statecodes` returns `Forbidden` also: <img alt="Screenshot 2023-03-20 at 2 42 48 PM" width="1710" src="https://user-images.githubusercontent.com/92752107/226448670-f6b63431-22e9-4102-a498-bd7f6634ebfc.png">
> > > It is `sslvpn.ncsa.illinois.edu` right?
> > 
> > 
> > I think that you have to use cisco anyconnect https://answers.uillinois.edu/illinois/page.php?id=98773
> 
> Ah, yes I am..here is the setting: <img alt="Screenshot 2023-03-20 at 2 48 54 PM" width="425" src="https://user-images.githubusercontent.com/92752107/226450014-80443c03-193a-4d36-afa7-220277c064e5.png">

which group are you in? Try ncsa-vpn-default

pengyin-shan reviewed (approved):
Tested and worked on my machine, good job!

I was to suggest using `dotenv` library and use `.env` file to control environmental variables. We can store the dev version of `.env` file on our dev machines, and a different version of `.env` file specifically on production server, then the app will look into this `.env` file for all environmental variables. 

All we need to do is to make `.env` on the production server safe, and we don't need to disrupt any machine's local environmental variable. What do you think?

ywkim312 commented:
> Tested and worked on my machine, good job!
> 
> I was to suggest using `dotenv` library and use `.env` file to control environmental variables. We can store the dev version of `.env` file on our dev machines, and a different version of `.env` file specifically on production server, then the app will look into this `.env` file for all environmental variables.
> 
> All we need to do is to make `.env` on the production server safe, and we don't need to disrupt any machine's local environmental variable. What do you think?

yes dot env is a good idea. We will consider that. Currently, we are using the docker compose and we set up some env variables in yaml files so making .env file for not very many env variables might not be necessary at this moment. If there are more variables, we will definitely need to consider using it.

pengyin-shan commented:
> > Tested and worked on my machine, good job!
> > I was to suggest using `dotenv` library and use `.env` file to control environmental variables. We can store the dev version of `.env` file on our dev machines, and a different version of `.env` file specifically on production server, then the app will look into this `.env` file for all environmental variables.
> > All we need to do is to make `.env` on the production server safe, and we don't need to disrupt any machine's local environmental variable. What do you think?
> 
> yes dot env is a good idea. We will consider that. Currently, we are using the docker compose and we set up some env variables in yaml files so making .env file for not very many env variables might not be necessary at this moment. If there are more variables, we will definitely need to consider using it.

yeah, if we feel the yaml file could create security threats or we need more environmental variables in the future, we should turn to dot env.

ywkim312 commented:
> > > Tested and worked on my machine, good job!
> > > I was to suggest using `dotenv` library and use `.env` file to control environmental variables. We can store the dev version of `.env` file on our dev machines, and a different version of `.env` file specifically on production server, then the app will look into this `.env` file for all environmental variables.
> > > All we need to do is to make `.env` on the production server safe, and we don't need to disrupt any machine's local environmental variable. What do you think?
> > 
> > 
> > yes dot env is a good idea. We will consider that. Currently, we are using the docker compose and we set up some env variables in yaml files so making .env file for not very many env variables might not be necessary at this moment. If there are more variables, we will definitely need to consider using it.
> 
> yeah, if we feel the yaml file could create security threats or we need more environmental variables in the future, we should turn to dot env.

yaml file doesn't have any security issue at this moment. It is just to see if it is "Development" or "production". So it should be  okay for now.

sandeep-ps commented on webpack.common.js:
```diff
@@ -94,6 +94,10 @@ module.exports = {
         }),
         new Webpack.DefinePlugin({
+            "process.env": {
+                // "NODE_ENV": JSON.stringify("development"),
```
Is the commented-out code needed? If not, could it be removed?

ywkim312 replied:
removed

sandeep-ps commented on src/utils/math.test.js:
```diff
@@ -2,6 +2,7 @@ import { sum } from './math';
 
 describe('sum function should', () => {
     test('return sum of an array of numbers', () => {
+        // eslint-disable-next-line no-undef
```
Similar thing here. Is the commented-out code needed? If not, could it be removed?

ywkim312 replied:
removed

sandeep-ps replied:
Thanks.

sandeep-ps reviewed (commented):
Could we remove the raw JSON file for all programs for the source code? 

Please update the CHANGELOG.

ywkim312 commented:
> The application seems to be working fine. Thanks.
> 
> Could we remove the raw JSON file for all programs for the source code?
> 
> Please update the CHANGELOG.

removed the folder and updated changelog

sandeep-ps commented:
Hi, @ywkim312, I have a few questions at this time. Is the goal of the PR to replace all JSON files or only a few? Let's say this PR aims to replace a single JSON file with API calls across all pages (which is perfectly fine),  I can review based on that. Or, if the goal is to replace all JSON files across all pages, I can continue my review based on that. Could you please clarify?

sandeep-ps commented:
Okay. Thanks. I will continue with the review soon.

sandeep-ps commented on .gitignore:
```diff
@@ -5,6 +5,10 @@ build/
 
+# IDEA:
+.idea/
+**.iml
```
```suggestion
```

ywkim312 replied:
removed

sandeep-ps commented on src/components/AllProgramMap.tsx:
```diff
@@ -176,10 +182,31 @@ MapChart.propTypes = {
+    const [allStatesData, setAllStatesData] = useState([]);
+
+    useEffect(() => {
+        const statecode_url = `${config.apiUrl}/statecodes`;
```
I see that `/statecodes` and `/states` endpoints are in this PR. Do you think it should be removed?

pengyin-shan replied:
is these state codes define full state names and abbreviations? If so, one json like this would be helpful (I used one for the SNAP table), but we might not want to have it as part of the API

ywkim312 replied:
both endpoints are needed. The reason why the json file are still there is because, if I remove those files, the other maps will not work.

sandeep-ps replied:
Okay. Got it. Resolving this comment.

sandeep-ps commented on src/data/allstates.json:
Please see the diff in this file. There is some random text. I think this JSON file should not be updated in this PR.

ywkim312 replied:
hmmm weird. I had no intension to update those files at all. Maybe something happened when merge conflict get fixed or I don't know why it happened. Anyway, reverted it back

sandeep-ps replied:
Okay. I also thought it would have crept in unknowingly. Thanks.

sandeep-ps commented on src/data/stateCodes.json:
Same here as well? stateCodes.json shouldn't be updated in this PR, right?

ywkim312 replied:
same here. No intension to update this file. Reverted

sandeep-ps commented:
There are docker build errors and other GitHub action failures. Those need to be addressed, too, right?

ywkim312 commented on .gitignore:
```diff
@@ -5,6 +5,10 @@ build/
 .env
-*.log
-*.DS_Store
+.huskyrc*.DS_Store
```
modified

ywkim312 commented:
> There are docker build errors and other GitHub action failures. Those need to be addressed, too, right?

this is also weird.... it worked okay when I first created PR. maybe some other things happened, maybe during the merging to develop to fix conflict... maybe not. I don't know why...

> There are docker build errors and other GitHub action failures. Those need to be addressed, too, right?

I think that it happened after it got merged to develop to fix the conflict. It complains about /home/runner/work/pdl-frontend/pdl-frontend/src/utils/math.test.js file. I will try to remove it since the other PR removed it.... That is why I don't like to make a PR sit there too long.

sandeep-ps commented on src/utils/math.test.js:
```diff
@@ -2,7 +2,6 @@ import { sum } from './math';
 
 describe('sum function should', () => {
     test('return sum of an array of numbers', () => {
-        // eslint-disable-next-line no-undef
```
I think if you re-instate this, the build error related to math.text.js will go away. I'm sorry, I thought it was a regular comment. I should have looked closer.

sandeep-ps commented:
It looks like allPrograms.json is used directly in LandingPageMap.tsx and LandingPageMapTab.tsx. My local build error was because of that. So, either the API call should replace those references, or the file might need to be uploaded back. :-(

sandeep-ps commented:
In Wendy's PR, she removed both math-related files in tests. I think they are related. I guess you could either add it back and uncomment this line mentioned here: https://github.com/policy-design-lab/pdl-frontend/pull/21#discussion_r1153644565 or you could try removing the other file too.

pengyin-shan commented:
> In Wendy's PR, she removed both math-related files in tests. I think they are related. I guess you could either add it back and uncomment this line mentioned here: [#21 (comment)](https://github.com/policy-design-lab/pdl-frontend/pull/21#discussion_r1153644565) or you could try removing the other file too.

@ywkim312 I don't know if this will be helpful to your PR, but if you are going to remove both math-related files (in the `utils` folder, one is function and another one is the corresponding test), make sure you change the test command in `package.json` to `"test": "jest --coverage --passWithNoTests" so you won't get the error reported in Github test.

ywkim312 commented:
@pengyin-shan @sandeep-ps Thank you for your help. Now all the test github action ran thought without error after adding `passWithNoTests` flag and no math files.

pengyin-shan commented:
> @pengyin-shan @sandeep-ps Thank you for your help. Now all the test github action ran thought without error after adding `passWithNoTests` flag and no math files.

Great! the change is to make sure that _jest_ won't complain if there's no test.

sandeep-ps commented:
Thanks. I'm merging this PR.
